text/0000-builder-key-in-project-descriptor.md

```diff
@@ -0,0 +1,74 @@
+# Meta
+[meta]: #meta
+- Name: Builder key in project descriptor
+- Start Date: 2021-02-11
+- Author(s): @wburningham
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a new `builder` key in the `project.toml` file.
+
+# Definitions
+[definitions]: #definitions
+
+* _project descriptor_ - The `project.toml` file that describes a project.
+
+# Motivation
+[motivation]: #motivation
+
+New optional `builder` field in the `project.toml` would allow configuration via file rather than CLI flags. This is useful when teams/companies have scaffolding/templetazing tools to help developers initialize a new project (ex: an internal tool/system can generate a `project.toml` with a default custom builder that matches the project). It is expected that a project/repo can store/serialize a `builder` in a parsable format (ie the `project.toml` rather than a `--builder` flag in a bash script). Being able to search repos to identify consumers of a specifier builder is helpful for auditing and communciating changes (hard to do when the `--builder` flag is specified in an unknown script/location).
```

Comment:
I quite like this proposal! Curious about a possible edge-case - How would this key interact with a platform if someone explicitly specifies a different builder.

For example `pack build image-name --builder another-builder -d project.toml`. Would it error/refuse to build, or would it be overridden? I guess I am also curious what the existing expected behaviour for the same with `buildpacks` key vs providing an explicit flag via the command line that contradicts with it. and https://github.com/buildpacks/rfcs/blob/main/text/0019-project-descriptor.md.

Reply:
I would assume that an explicit CLI flag value should override a project descriptor. That's what we do with run-mirrors, for instance

Reply:
That's good to know!


Probably off topic for this PR but - is this documented somewhere in the spec/rfc? If not, might be a good idea to enforce consistent behavior with descriptor overrides.

Reply:
Definitely agree that it should be added in the RFC here, at least. :D

Reply:
@wburningham could we add the proposed order of precedence to this RFC?

My thinking would be that this is the desired order of precedence:

1. `--builder` 👈 highest precedence
2. `project.toml` → `build.builder` 
3. configured builder (set via `pack config default-builder`)  👈 lowest precedence

Reply:
I've documented the order of precedence in my latest commit.